File: java/smartlist/src/main/java/com/smartlist/module/candidate/CandidateController.java
package com.smartlist.module.candidate;

import com.smartlist.module.auth.dto.response.AuthRespDTO;
import com.smartlist.module.candidate.dto.request.CreateCandidateReqDTO;
import com.smartlist.module.candidate.dto.request.SubmitAnswerReqDTO;
import com.smartlist.module.candidate.dto.request.UpdateCandidateProfileReqDTO;
import com.smartlist.module.candidate.dto.response.CandidateAssessmentIntroRespDTO;
import com.smartlist.module.candidate.dto.response.CandidateAssessmentRespDTO;
import com.smartlist.module.candidate.dto.response.CandidateRespDTO;
import com.smartlist.module.candidate.dto.response.CandidateResponseRespDTO;
import com.smartlist.utils.AppResponseCodes;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Positive;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.annotation.Secured;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@Tag(name = "Candidate", description = "Candidate API")
@RequestMapping(value ="", produces = { "application/json" })
@ResponseStatus(HttpStatus.OK)
@Validated
@Secured({"CANDIDATE"})
public class CandidateController {

    private static final String ROOT_ROUTE = "/candidate";

    private final CandidateService candidateService;

    @Autowired
    CandidateController(final CandidateService candidateService){
        this.candidateService = candidateService;
    }

    @Operation(summary = "Get assessment introduction by assessment token")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = CandidateAssessmentIntroRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ASSESSMENT_TOKEN_DESCRIPTION, content = @Content)
    })
    @GetMapping(CandidateController.ROOT_ROUTE + "/assessment/token/{token}/intro")
    public CandidateAssessmentIntroRespDTO getCandidateAssessmentIntro(@PathVariable("token") @NotBlank String token){
        return this.candidateService.getCandidateAssessmentIntro(token);
    }

    @Operation(summary = "Create candidate")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = Integer.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.CANDIDATE_EXISTS_DESCRIPTION, content = @Content)
    })
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(CandidateController.ROOT_ROUTE)
    public Integer create(@RequestBody() @Valid CreateCandidateReqDTO createCandidateReqDto) {
        return this.candidateService.createCandidate(createCandidateReqDto);
    }


    @Operation(summary = "Send OTP")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.CANDIDATE_ALREADY_VERIFIED_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping(CandidateController.ROOT_ROUTE + "/{id}/send-verification-code")
    public void resendOTP(@PathVariable("id") @Positive Integer id) {
        this.candidateService.sendCandidateOTP(id);
    }

    @Operation(summary = "Update candidate profile")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_PHOTO_FILE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_VIDEO_FILE_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping(CandidateController.ROOT_ROUTE + "/{id}/profile")
    public void updateProfile(@PathVariable("id") @Positive Integer id, @RequestBody() @Valid UpdateCandidateProfileReqDTO updateCandidateProfileReqDto) {
        this.candidateService.updateCandidateProfile(id, updateCandidateProfileReqDto);
    }

    @Operation(summary = "Get candidate by contact number")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = CandidateRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_CONTACT_NUMBER_DESCRIPTION, content = @Content)
    })
    @GetMapping(CandidateController.ROOT_ROUTE + "/contact-number/{contactNumber}")
    public CandidateRespDTO getByContactNumber(@PathVariable("contactNumber") @NotBlank String contactNumber) {
        return this.candidateService.getCandidateByContactNumber(contactNumber);
    }

    @Secured({"SUPERADMIN", "ADMIN", "MEMBER", "CANDIDATE"})
    @Operation(summary = "Get candidate by Id")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = CandidateRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping(CandidateController.ROOT_ROUTE + "/{id}")
    public CandidateRespDTO get(@PathVariable("id") @Positive Integer id){
        return  this.candidateService.getCandidateById(id);
    }

    @Operation(summary = "Create candidate assessment")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = Integer.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.CANDIDATE_ASSESSMENT_EXISTS_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.ASSESSMENT_ACTION_DENIED_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ASSESSMENT_TOKEN_DESCRIPTION, content = @Content)

    })
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(CandidateController.ROOT_ROUTE + "/{id}/assessment/token/{token}")
    public Integer createAssessment(@PathVariable("id") @Positive Integer id, @PathVariable("token") @NotEmpty String token) {
        return this.candidateService.createCandidateAssessment(id, token);
    }

    @Secured({"ADMIN", "MEMBER"})
    @Operation(summary = "Search candidates by name and organization")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = CandidateRespDTO.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ORGANIZATION_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("/organization/{organizationId}" + CandidateController.ROOT_ROUTE)
    public List<CandidateRespDTO> search(@PathVariable("organizationId") @Positive Integer organizationId, @RequestParam("name") @NotBlank String name) {
        return this.candidateService.getCandidateByNameForOrganization(organizationId, name);
    }

    @Operation(summary = "Get candidate assessment by assessment token")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = CandidateAssessmentRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.CANDIDATE_ASSESSMENT_TIMED_OUT_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ASSESSMENT_TOKEN_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS_DESCRIPTION, content = @Content)
    })
    @GetMapping(CandidateController.ROOT_ROUTE + "/{id}/assessment/token/{token}")
    public CandidateAssessmentRespDTO getCandidateAssessmentByToken(@PathVariable("id") @Positive Integer id, @PathVariable("token") @NotBlank String token) {
        return this.candidateService.getCandidateAssessmentByToken(id, token);
    }

    @Secured({"SUPERADMIN", "ADMIN", "MEMBER", "CANDIDATE"})
    @Operation(summary = "Get candidate assessment by assessment id")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = CandidateAssessmentRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS_DESCRIPTION, content = @Content)
    })
    @GetMapping(CandidateController.ROOT_ROUTE + "/{id}/assessment/{assessmentId}")
    public CandidateAssessmentRespDTO getCandidateAssessmentByAssessmentId(
            @PathVariable("id") @Positive Integer id,
            @PathVariable("assessmentId") @Positive Integer assessmentId){
        return this.candidateService.getCandidateAssessmentByAssessmentId(id, assessmentId);
    }


    @Operation(summary = "Log candidate assessment attempt")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(CandidateController.ROOT_ROUTE + "/{id}/assessment/{assessmentId}/attempt/log")
    public void logAssessmentAttempt(@PathVariable("id") @Positive Integer id, @PathVariable("assessmentId") @Positive Integer assessmentId) {
        this.candidateService.logAssessmentAttempt(id, assessmentId);
    }

    @Operation(summary = "Submit answer")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.CANDIDATE_ASSESSMENT_ACTION_DENIED_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = CandidateResponseCodes.ANSWER_RESPONSE_INCORRECT_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ASSESSMENT_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_QUESTION_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ANSWER_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_VIDEO_FILE_ID_DESCRIPTION, content = @Content)
    })
    @PutMapping(CandidateController.ROOT_ROUTE + "/candidate-assessment/{candidateAssessmentId}/question/{questionId}/submit-answer")
    public void submitAnswer(
            @PathVariable("candidateAssessmentId") @Positive Integer candidateAssessmentId,
            @PathVariable("questionId") @Positive Integer questionId,
            @RequestBody() @Valid SubmitAnswerReqDTO submitAnswerReqDto) {
       this.candidateService.submitAnswer(candidateAssessmentId, questionId, submitAnswerReqDto);
    }

    @Operation(summary = "Set active assessment block")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ASSESSMENT_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping(CandidateController.ROOT_ROUTE + "/candidate-assessment/{candidateAssessmentId}/assessment-block/{assessmentBlockId}/active")
    public void setActiveAssessmentBlock(
            @PathVariable("candidateAssessmentId") @Positive Integer candidateAssessmentId,
            @PathVariable("assessmentBlockId") @Positive Integer assessmentBlockId) {
        this.candidateService.setActiveAssessmentBlock(candidateAssessmentId, assessmentBlockId);
    }

    @Operation(summary = "Mark assessment complete")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping(CandidateController.ROOT_ROUTE + "/candidate-assessment/{candidateAssessmentId}/complete")
    public void markAssessmentComplete(@PathVariable("candidateAssessmentId") @Positive Integer candidateAssessmentId) {
        this.candidateService.markCandidateAssessmentComplete(candidateAssessmentId);
    }

    @Secured({"ADMIN", "MEMBER", "CANDIDATE"})
    @Operation(summary = "Get candidate responses for question")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = CandidateResponseRespDTO.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_CANDIDATE_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = CandidateResponseCodes.INVALID_QUESTION_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping(CandidateController.ROOT_ROUTE + "/candidate-assessment/{candidateAssessmentId}/question/{questionId}/candidate-response")
    public List<CandidateResponseRespDTO> getResponsesForQuestion(
            @PathVariable("candidateAssessmentId") @Positive Integer candidateAssessmentId,
            @PathVariable("questionId") @Positive Integer questionId) {
        return  this.candidateService.getResponsesByQuestionId(candidateAssessmentId, questionId);
    }


}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/candidate/candidate.controller.ts
import { ParseIntPipe, Patch, Put, Query } from '@nestjs/common';
import { Controller, Get, Post, Body, Param, HttpCode } from '@nestjs/common';
import {
  ApiOperation,
  ApiProduces,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { Role } from 'src/enums/role';
import { Roles } from 'src/guards/roles.decorator';
import { AuthRespDto } from '../auth/dto/response/auth.resp.dto';
import { Public } from '../auth/jwt-auth.guard';
import { CandidateResponseCodes } from './candidate.response.codes';
import { CandidateService } from './candidate.service';
import { CreateCandidateReqDto } from './dto/request/create-candidate.req.dto';
import { SubmitAnswerReqDto } from './dto/request/submit-answer.req.dto';
import { UpdateCandidateProfileReqDto } from './dto/request/update-candidate.req.dto';
import { CandidateAssessmentRespDto } from './dto/response/candidate-assessment.resp.dto';
import { CandidateResponseRespDto } from './dto/response/candidate-response.resp.dto';
import { CandidateRespDto } from './dto/response/candidate.resp.dto';
import { AuthService } from '../auth/auth.service';
import { Candidate } from 'src/entities/candidate.entity';
import { CandidateAssessmentIntroRespDto } from './dto/response/candidate-assessment-intro.resp.dto';

@ApiTags('Candidate')
@Controller('')
@Roles(Role.CANDIDATE)
export class CandidateController {
  static rootRoute = 'candidate/';

  constructor(
    private readonly candidateService: CandidateService,
    private readonly authService: AuthService,
  ) {}

  @ApiOperation({ summary: 'Get assessment introduction by assessment token' })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: CandidateAssessmentIntroRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_ASSESSMENT_TOKEN)
  @Public()
  @HttpCode(200)
  @Get(CandidateController.rootRoute + 'assessment/token/:token/intro')
  async getCandidateAssessmentIntro(
    @Param('token') token: string,
  ): Promise<CandidateAssessmentIntroRespDto> {
    return await this.candidateService.getCandidateAssessmentIntro(token);
  }

  @ApiOperation({ summary: 'Create candidate' })
  @ApiResponse({ ...CandidateResponseCodes.CREATED, type: Number })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_EXISTS)
  @ApiProduces('application/text')
  @Public()
  @HttpCode(201)
  @Post(CandidateController.rootRoute)
  async create(
    @Body() createCandidateReqDto: CreateCandidateReqDto,
  ): Promise<number> {
    return await this.candidateService.createCandidate(createCandidateReqDto);
  }

  @ApiOperation({ summary: 'Verify OTP' })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: AuthRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_VERIFICATION_CODE)
  @Public()
  @HttpCode(200)
  @Patch(CandidateController.rootRoute + ':id/verify/:otp')
  async verifyOTP(
    @Param('id', ParseIntPipe) id: number,
    @Param('otp', ParseIntPipe) otp: number,
  ): Promise<AuthRespDto> {
    const candidate: Candidate = await this.candidateService.verifyCandidateOTP(
      id,
      otp,
    );
    return await this.authService.loginCandidate(candidate);
  }

  @ApiOperation({ summary: 'Send OTP' })
  @ApiResponse({ ...CandidateResponseCodes.SUCCESS })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ALREADY_VERIFIED)
  @Public()
  @HttpCode(200)
  @Patch(CandidateController.rootRoute + ':id/send-verification-code')
  async resendOTP(@Param('id', ParseIntPipe) id: number) {
    return await this.candidateService.sendCandidateOTP(id);
  }

  @ApiOperation({ summary: 'Get candidate by contact number' })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: CandidateRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_CONTACT_NUMBER)
  @Public()
  @HttpCode(200)
  @Get(CandidateController.rootRoute + 'contact-number/:contactNumber')
  async getByContactNumber(
    @Param('contactNumber') contactNumber: string,
  ): Promise<CandidateRespDto> {
    return await this.candidateService.getCandidateByContactNumber(
      contactNumber,
    );
  }

  @ApiOperation({ summary: 'Get candidate by Id' })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: CandidateRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @Roles(Role.CANDIDATE, Role.SUPERADMIN, Role.ADMIN, Role.MEMBER)
  @HttpCode(200)
  @Get(CandidateController.rootRoute + ':id')
  async get(@Param('id', ParseIntPipe) id: number): Promise<CandidateRespDto> {
    return await this.candidateService.getCandidateById(id);
  }

  @ApiOperation({
    summary: 'Search candidates by name and organization',
  })
  @ApiResponse(CandidateResponseCodes.SUCCESS)
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_ORGANIZATION_ID)
  @Roles(Role.ADMIN, Role.MEMBER)
  @HttpCode(200)
  @Get('organization/:organizationId/' + CandidateController.rootRoute)
  async search(
    @Param('organizationId', ParseIntPipe) organizationId: number,
    @Query('name') name: string,
  ) {
    return await this.candidateService.searchByNameForOrganization(
      organizationId,
      name,
    );
  }

  @ApiOperation({ summary: 'Get candidate assessment by assessment token' })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: CandidateAssessmentRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_ASSESSMENT_TOKEN)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_TIMED_OUT)
  @HttpCode(200)
  @Get(CandidateController.rootRoute + ':id/assessment/token/:token')
  async getCandidateAssessmentByToken(
    @Param('id', ParseIntPipe) id: number,
    @Param('token') token: string,
  ): Promise<CandidateAssessmentRespDto> {
    return await this.candidateService.getCandidateAssessmentByToken(id, token);
  }

  @ApiOperation({ summary: 'Get candidate assessment by assessment id' })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: CandidateAssessmentRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS)
  @Roles(Role.CANDIDATE, Role.SUPERADMIN, Role.ADMIN, Role.MEMBER)
  @HttpCode(200)
  @Get(CandidateController.rootRoute + ':id/assessment/:assessmentId')
  async getCandidateAssessmentByAssessmentId(
    @Param('id', ParseIntPipe) id: number,
    @Param('assessmentId', ParseIntPipe) assessmentId: number,
  ): Promise<CandidateAssessmentRespDto> {
    return await this.candidateService.getCandidateAssessmentByAssessmentId(
      id,
      assessmentId,
    );
  }

  @ApiOperation({ summary: 'Create candidate assessment' })
  @ApiResponse({ ...CandidateResponseCodes.CREATED, type: Number })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_EXISTS)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_ASSESSMENT_TOKEN)
  @ApiResponse(CandidateResponseCodes.ASSESSMENT_ACTION_DENIED)
  @ApiProduces('application/text')
  @HttpCode(201)
  @Post(CandidateController.rootRoute + ':id/assessment/token/:token')
  async createAssessment(
    @Param('id', ParseIntPipe) id: number,
    @Param('token') token: string,
  ): Promise<number> {
    return await this.candidateService.createCandidateAssessment(id, token);
  }

  @ApiOperation({ summary: 'Log candidate assessment attempt' })
  @ApiResponse({ ...CandidateResponseCodes.CREATED, type: Number })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiProduces('application/text')
  @HttpCode(201)
  @Post(
    CandidateController.rootRoute + ':id/assessment/:assessmentId/attempt/log',
  )
  async logAssessmentAttempt(
    @Param('id', ParseIntPipe) id: number,
    @Param('assessmentId') assessmentId: number,
  ): Promise<void> {
    return await this.candidateService.logAssessmentAttempt(id, assessmentId);
  }

  @ApiOperation({ summary: 'Update candidate profile' })
  @ApiResponse({ ...CandidateResponseCodes.SUCCESS })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_PHOTO_FILE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_VIDEO_FILE_ID)
  @HttpCode(200)
  @Patch(CandidateController.rootRoute + ':id/profile')
  async updateProfile(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateCandidateProfileReqDto: UpdateCandidateProfileReqDto,
  ) {
    return await this.candidateService.updateCandidateProfile(
      id,
      updateCandidateProfileReqDto,
    );
  }

  @ApiOperation({ summary: 'Submit answer' })
  @ApiResponse(CandidateResponseCodes.SUCCESS)
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ASSESSMENT_ID)
  @ApiResponse(CandidateResponseCodes.CANDIDATE_ASSESSMENT_ACTION_DENIED)
  @ApiResponse(CandidateResponseCodes.INVALID_QUESTION_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_ANSWER_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_VIDEO_FILE_ID)
  @ApiResponse(CandidateResponseCodes.ANSWER_RESPONSE_INCORRECT)
  @HttpCode(200)
  @Put(
    CandidateController.rootRoute +
      'candidate-assessment/:candidateAssessmentId/question/:questionId/submit-answer',
  )
  async submitAnswer(
    @Param('candidateAssessmentId', ParseIntPipe) candidateAssessmentId: number,
    @Param('questionId', ParseIntPipe) questionId: number,
    @Body() submitAnswerReqDto: SubmitAnswerReqDto,
  ) {
    return await this.candidateService.submitAnswer(
      candidateAssessmentId,
      questionId,
      submitAnswerReqDto,
    );
  }

  @ApiOperation({ summary: 'Set active assessment block' })
  @ApiResponse(CandidateResponseCodes.SUCCESS)
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ASSESSMENT_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_ASSESSMENT_BLOCK_ID)
  @HttpCode(200)
  @Patch(
    CandidateController.rootRoute +
      'candidate-assessment/:candidateAssessmentId/assessment-block/:assessmentBlockId/active',
  )
  async setActiveAssessmentBlock(
    @Param('candidateAssessmentId', ParseIntPipe) candidateAssessmentId: number,
    @Param('assessmentBlockId', ParseIntPipe) assessmentBlockId: number,
  ) {
    return await this.candidateService.setActiveAssessmentBlock(
      candidateAssessmentId,
      assessmentBlockId,
    );
  }

  @ApiOperation({ summary: 'Mark assessment complete' })
  @ApiResponse(CandidateResponseCodes.SUCCESS)
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ASSESSMENT_ID)
  @HttpCode(200)
  @Patch(
    CandidateController.rootRoute +
      'candidate-assessment/:candidateAssessmentId/complete',
  )
  async markAssessmentComplete(
    @Param('candidateAssessmentId', ParseIntPipe) candidateAssessmentId: number,
  ) {
    return await this.candidateService.markCandidateAssessmentComplete(
      candidateAssessmentId,
    );
  }

  @ApiOperation({
    summary: 'Get candidate responses for question',
  })
  @ApiResponse({
    ...CandidateResponseCodes.SUCCESS,
    type: CandidateResponseRespDto,
  })
  @ApiResponse(CandidateResponseCodes.BAD_REQUEST)
  @ApiResponse(CandidateResponseCodes.INVALID_CANDIDATE_ID)
  @ApiResponse(CandidateResponseCodes.INVALID_QUESTION_ID)
  @Roles(Role.ADMIN, Role.MEMBER, Role.CANDIDATE)
  @HttpCode(200)
  @Get(
    CandidateController.rootRoute +
      'candidate-assessment/:candidateAssessmentId/question/:questionId/candidate-response',
  )
  async getResponsesForQuestion(
    @Param('candidateAssessmentId', ParseIntPipe) candidateAssessmentId: number,
    @Param('questionId', ParseIntPipe) questionId: number,
  ): Promise<CandidateResponseRespDto[]> {
    return await this.candidateService.getResponsesByQuestionId(
      candidateAssessmentId,
      questionId,
    );
  }
}
